BD-02 BD-03 Add support for LTI embeds in course tabs and elsewhere

pcockwell opened the pull request:
Separates out the LTI 1.1 functionality from XBlock usage so that LTI embeds can
be added without XBlock usage in the rest of the LMS. Includes a helper method
used to generate the HTML and JavaScript content used to launch the LTI embed

**JIRA tickets**:
- [OSPR-4686](https://openedx.atlassian.net/browse/OSPR-4686)
- BD-02
- BD-03

**Dependencies**: None

**Merge deadline**: "None"

**Testing instructions**:

1. Add `open-craft` as a remote
```
git remote add open-craft git@github.com:open-craft/xblock-lti-consumer.git
``` 
2. Checkout this branch
```
git fetch open-craft
git checkout open-craft:patrick/BB-2523-add-support-for-lti-embeds-in-the-course-tabs-a
```
3. Follow install directions located in the repository [README](https://github.com/edx/xblock-lti-consumer/blob/master/README.rst)
4. Add a view to your `edx-platform` code that renders the `lti_embed`
```python
# edx-platform/lms/djangoapps/lti_embed/views.py
from django.contrib.auth.decorators import login_required
from django.http import HttpResponse

from lti_consumer.lti_1p1.contrib.django import lti_embed


@login_required
def basic_lti_embed(request):
    """
    Provides the LTI Embed outside of an xblock
    """

    return HttpResponse(lti_embed(
        html_element_id='direct-embed',
        lti_launch_url='http://lti.tools/saltire/tp',
        oauth_key='jisc.ac.uk',
        oauth_secret='secret',
        resource_link_id='unique-resource-link-id',
        user_id='student',
        roles='Student',
        context_id='some-page',
        context_title='Some page title',
        context_label='Some page label',
        result_sourcedid='unique-result-sourcedid',
        person_sourcedid=None,
        person_contact_email_primary=None,
        outcome_service_url=None,
        launch_presentation_locale=None
    ))
```
5. Add a `urls.py` file to the new `lti_embed` subdirectory that will render the above view
```python
# edx-platform/lms/djangoapps/lti_embed/urls.py
"""
URLs for the LTI Embedding test
"""

from django.conf.urls import url

from lti_embed.views import basic_lti_embed

urlpatterns = [
    url(r'^$', basic_lti_embed, name='lti_embed'),
]
```
5. Include your `lms/djangoapps/lti_embed/urls.py` file in the base `lms/urls.py` file
```python
# edx-platform/lms/urls.py
...
urlpatterns = [
    url(r'^$', branding_views.index, name='root'),  # Main marketing page, or redirect to courseware

    # ADD ONLY THIS LINE
    url(r'^lti_embed/', include('lti_embed.urls')),

    url(r'', include('student.urls')),
    # TODO: Move lms specific student views out of common code
    url(r'^dashboard/?$', student_views.student_dashboard, name='dashboard'),
    url(r'^change_enrollment$', student_views.change_enrollment, name='change_enrollment'),
    ...
]
```
6. Load your localhost LMS in your browser and login
7. Navigate to `localhost:18000/lti_embed` and your browser should generate the
LTI Launch and redirect you to `http://lti.tools/saltire/tp`
8. If you wish to remain on the LMS, you should wrap the template returned from
the `lti_embed` method with an `<iframe>`.

**Author notes and concerns**:

1. The `lti_embed` method uses `xblockutils.resources.ResourceLoader` to load the HTML
template used. The rationale for this is that `ResourceLoader` is agnostic
to XBlock code and functionality. It is recommended that this remain in use
until LTI1.3 support is merged, or a better means of loading the template is
made available.

**Reviewers**
- [ ] @giovannicimolin
- [ ] edX reviewer[s] TBD

edx-webhook commented:
Thanks for the pull request, @pcockwell! I've created [OSPR-4686](https://openedx.atlassian.net/browse/OSPR-4686) to keep track of it in JIRA, where we prioritize reviews.  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

Please let us know once your PR is ready for our review and all tests are green.

xitij2000 commented on lti_consumer/lti_1p1/exceptions.py:
```diff
@@ -0,0 +1,9 @@
+class LtiError(Exception):
+    """
+    General error class for LTI Consumer usage.
+    """
```
I think this should be a common error, and this particular error can be Lti1p1Error?

pcockwell replied:
I'm not quite sure I follow what you mean by `a common error`? Should I be moving `LtiError` into a separate file, or just removing it altogether and replacing it with `Lti1p1Error`?

giovannicimolin replied:
@pcockwell I think it makes more sense to rename this one to `Lti1p1Error` for LTI 1.1 specific cases, and use a generic LTI error when throwing from `lti_consumer.py` (the block).

pcockwell replied:
@giovannicimolin I've pushed some changes. Please let me know if this was your intended result.

xitij2000 commented on lti_consumer/lti.py:
```diff
@@ -7,91 +7,19 @@
-    return score, json_obj.get('comment', "")
-
-
 class LtiConsumer(object):  # pylint: disable=bad-option-value, useless-object-inheritance
```
This should be removed now right?

pcockwell replied:
More than happy to remove it. From what I can tell, it is no longer used anywhere.

pcockwell replied:
Resolved in most recent commit

xitij2000 commented on lti_consumer/lti_consumer.py:
```diff
@@ -425,6 +393,21 @@ class LtiConsumerXBlock(StudioEditableXBlockMixin, XBlock):
+        help=_("User's email; only used if ask_to_send_email is True, and user authorizes it."),
+        default=None,
+        scope=Scope.user_state
+    )
```
I think this has PII concerns. I'm not sure you should be storing this in the XBlock.

pcockwell replied:
Resolved in most recent commit

xitij2000 commented on lti_consumer/lti_consumer.py:
```diff
@@ -723,6 +718,32 @@ def is_past_due(self):
+        key, secret = self.lti_provider_key_secret
+        return LtiConsumer1p1(self.launch_url, key, secret)
+
+    def extract_real_user_data(self):
+        """
+        Set actual user data on xblock from the runtime
+        """
+        if callable(self.runtime.get_real_user):
+            real_user_object = self.runtime.get_real_user(self.runtime.anonymous_student_id)
+            self.user_email = getattr(real_user_object, "email", "")
+            self.user_username = getattr(real_user_object, "username", "")
+            user_preferences = getattr(real_user_object, "preferences", None)
+
+            if user_preferences is not None:
+                language_preference = user_preferences.filter(key='pref-lang')
+                if len(language_preference) == 1:
+                    self.user_language = language_preference[0].value
```
Instead of storing these in self, just return them from `extract_real_user_data`?

```suggestion
    def extract_real_user_data(self):
        """
        Set actual user data on xblock from the runtime
        """
        user_email = None
        user_username = None
        user_language = None
        if callable(self.runtime.get_real_user):
            real_user_object = self.runtime.get_real_user(self.runtime.anonymous_student_id)
            user_email = getattr(real_user_object, "email", "")
            user_username = getattr(real_user_object, "username", "")
            user_preferences = getattr(real_user_object, "preferences", None)

            if user_preferences is not None:
                language_preference = user_preferences.filter(key='pref-lang')
                if len(language_preference) == 1:
                    user_language = language_preference[0].value
        return user_email, user_username, user_lanauge
```

pcockwell replied:
I've implemented something similar to this in the most recent commit

natabene commented:
@pcockwell Thank you for your contribution. Is this a BD PR by any chance?

xitij2000 commented:
@natabene Yes, it is. Will mark it in the ticket title.

giovannicimolin commented:
@davidjoy @davestgermain Since this affects the LTI project merge, can you take a look and do a review pass too?

davidjoy commented on lti_consumer/lti_1p1/contrib/django.py:
```diff
@@ -0,0 +1,126 @@
+    IMPORTANT NOTE: This method uses keyword only arguments as described in PEP 3102.
+    Given the large number of arguments for this method, there is a  desire to
+    guarantee that developers using this method know which arguments are being set
+    to which values. This syntax is NOT backwards compatible with python 2.X, but is
```
Do you think this should be a breaking change/result in a major version bump for the repo?

davidjoy commented on lti_consumer/lti_consumer.py:
```diff
@@ -709,6 +677,18 @@ def prefixed_custom_parameters(self):
 
                 custom_parameters[six.text_type(param_name)] = six.text_type(param_value)
+
+        custom_parameters[six.text_type('custom_component_display_name')] = six.text_type(self.display_name)
```
Do we need to be using six in this repo still?  My understanding was that it was for ensuring python 2-3 compatibility.  With the `lti_embed` syntax change above, we're no longer compatible with python 2 anyway, right?

pcockwell replied:
@davidjoy Happy to remove `six` if we are comfortable with the repo not being python 2 compatible anymore. In this particular case though, if people are only using the `LtiConsumerXBlock`, everything they have access to from that class is still in fact python 2 compatible. I'm not sure if we are comfortable with partial compatibility or not though.

I feel like the options we likely have available are:
- Remove `six`, bump the major version, and not longer support Python 2 compatibility
- Modify the `lti_embed` syntax to be Python 2 compatible
- Be comfortable with partial compatibility (and possibly make this a minor version bump instead of a patch version bump)

Please advise to your desired course of action. I think it probably makes sense for @giovannicimolin to weigh in on this as well

xitij2000 replied:
I think it makes sense to break compatibility and make a major release, here is why:
-This block comes pre-installed with the platform, and it's unclear if anyone upgrades it to a newer version. 
-This version will come with Koa, and I doubt Python 2 will be supported for that.
- Python 2 itself isn't supported anymore.

pcockwell replied:
Sounds good, I will bump the major version. `six` is used quite extensively in this repo though, so I'm inclined not to clutter this PR up with the removal of `six`. I will remove it from any code that was touched in this PR though.

davidjoy replied:
👍 On this.  Agree on the major version bump, especially after conferring with others on the platform team here.  Agree that removing six can happen separately.

pcockwell commented:
@giovannicimolin @xitij2000 @davidjoy @davestgermain 
Can I please get another round of reviews? Thanks

davestgermain commented on lti_consumer/lti_consumer.py:
```diff
@@ -762,8 +774,39 @@ def lti_launch_handler(self, request, suffix=''):  # pylint: disable=unused-argu
+
+        username = None
+        email = None
+        if self.ask_to_send_username and real_user_data['user_username']:
```
Since most installations will not be sending username and email, can you move `self.extract_real_user_data()` to be conditional based on `self.ask_to_send_username or self.ask_to_send_email`?

That will avoid looking up the user unless it's needed.

pcockwell replied:
Are we sure we want to do that given that `self.extract_real_user_data()` also returns the `user_language`, which is set independently of the values of `self.ask_to_send_username` and `self.ask_to_send_email`. Or do we not care about `user_language` if those values are both false?

Also, with regards to setting values for `username` and `email` from the result of `self.extract_real_user_data()`, if I use an `or` statement as you suggested, should I then subsequently check `self.ask_to_send_username` before setting the username, and `self.ask_to_send_email` before setting the email?

davestgermain replied:
Ok, I missed the `user_language`. And the old implementation was also doing the same thing you're doing. So, it's not easy to optimize this right now.

davestgermain commented on lti_consumer/lti_1p1/contrib/django.py:
```diff
@@ -0,0 +1,125 @@
+This module provides functionality for rendering an LTI embed without an XBlock.
+"""
+
+from __future__ import absolute_import, unicode_literals
```
These aren't needed for python 3

pcockwell replied:
This is resolved in the most recent commit

davestgermain commented on lti_consumer/lti_1p1/contrib/django.py:
```diff
@@ -0,0 +1,125 @@
+        {
+            key: value
+            for key, value in custom_parameters.items()
+            if key.startswith('custom_')
```
Why do the custom parameters have to start with `custom_`? Is that a requirement of the spec?

pcockwell replied:
I was following the specifications outlined [here](https://github.com/edx/xblock-lti-consumer/blob/98d4262c5d6dab7a183e029a48822bc9104e6022/lti_consumer/lti_consumer.py#L811-L841) in the original code.

The implementation guide for LTI1.1 is linked below, and if you do a page search for `custom_keyname`. I am having a bit of difficulty understanding whether or not the prefix is required, but it certainly seems to be recommended at the least.

https://www.imsglobal.org/specs/ltiv1p1/implementation-guide

davestgermain replied:
Thanks. That seems clear enough.

pcockwell commented on lti_consumer/lti_consumer.py:
```diff
@@ -708,7 +674,19 @@ def prefixed_custom_parameters(self):
+
+        if self.due:  # pylint: disable=no-member
+            custom_parameters.update({
+                str('custom_component_due_date'): str(self.due.strftime('%Y-%m-%d %H:%M:%S'))  # pylint: disable=no-member
```
Resolved in most recent commit

pcockwell commented on lti_consumer/lti_1p1/oauth.py:
```diff
@@ -117,7 +115,7 @@ def verify_oauth_body_signature(request, lti_provider_secret, service_url):
             request.body
         )
-        raise LtiError("OAuth body hash verification is failed.")
+        raise Lti1p1Error("OAuth body hash verification is failed.")
```
Resolved in most recent commit

pcockwell commented on lti_consumer/lti_1p1/consumer.py:
```diff
@@ -0,0 +1,395 @@
+            person_contact_email_primary (string):  Primary contact email address of the user
+        """
+        self.lti_user_data = {
+            str('user_id'): user_id,
```
Resolved in most recent commit

pcockwell commented on lti_consumer/lti_1p1/consumer.py:
```diff
@@ -0,0 +1,395 @@
+                raise Lti1p1Error(msg)
+        except (TypeError, ValueError) as err:
+            msg = "Could not convert resultScore to float: {}".format(str(err))
+            log.error("[LTI] %s", msg)
```
This is the only occurrence of this that I found reasonable to do this with as the other case where an exception was called, it was the actual exception message that was desired. Please let me know if you'd like anything else to be changed

pcockwell commented:
Also want to point out that there are multiple file conflicts since the merge of `Lti1p3` content. These conflicts are in direct contradiction of the desire to use the new `Lti1p1Consumer` class in this PR. Can you please advise as to how to proceed.

pcockwell commented:
I merged master in once again. While tests (locally) do pass, I noticed a significant amount of error messages such as this one:
```
Traceback (most recent call last):
  File "/edx/src/xblock-lti-consumer/venv/lib/python3.5/site-packages/django/template/base.py", line 829, in _resolve_lookup
    current = current[bit]
KeyError: 'max'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/edx/src/xblock-lti-consumer/venv/lib/python3.5/site-packages/django/template/base.py", line 837, in _resolve_lookup
    current = getattr(current, bit)
AttributeError: 'dict' object has no attribute 'max'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/edx/src/xblock-lti-consumer/venv/lib/python3.5/site-packages/django/template/base.py", line 843, in _resolve_lookup
    current = current[int(bit)]
ValueError: invalid literal for int() with base 10: 'max'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/edx/src/xblock-lti-consumer/venv/lib/python3.5/site-packages/django/template/base.py", line 850, in _resolve_lookup
    (bit, current))  # missing attribute
django.template.base.VariableDoesNotExist: Failed lookup for key [max] in {'has_list_values': False, 'default': 1.0, 'display_name': <Mock name='mock.service().ugettext()' id='139697843048064'>, 'list_values': None, 'name': 'weight', 'value': 1.0, 'has_values': False, 'is_set': False, 'type': 'float', 'min': 0, 'allow_reset': True, 'help': <Mock name='mock.service().ugettext()' id='139697843048064'>}
```

I switched to the master branch and noticed these messages do still appear while on master, so it isn't any code I've introduced. These errors appear for `max`, `min` and `step` almost identically.

@davestgermain @giovannicimolin Could I get another review pass on this please?

giovannicimolin commented:
@pcockwell The changes look good, and the issues don't seem to be related to your changes.
I think we need a new version bump for this (LTI 1.3 implementation already bumped this up to `2.0.1.1`).

I've checked that the LTI embed is still working with the latest changes, so this is good to go for me. :+1: 

Can you squash the commits and address the open discussions above?

davestgermain commented:
@pcockwell I don't know why those django template errors are appearing. We aren't using django templates in this repo, as far as I can tell.

So, just squash commits and increase the revision, and this'll be good to go.

pcockwell commented:
@davestgermain Changes have been squashed into a single commit, and the revision has been bumped to 2.0.2

edx-webhook commented:
@pcockwell 🎉 Your pull request was merged!

Please take a moment to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=edx/xblock-lti-consumer&entry.867055334=https://github.com/edx/xblock-lti-consumer/pull/77&entry.1484655318=https://github.com/pcockwell&entry.752974735=2020-06-11+03:21&entry.1917517419=2020-07-02+13:45&entry.2133058324=Yes) so we can improve your experience in the future.
